qt/Torrent.cc

```diff
@@ -173,228 +108,262 @@ Torrent::KeyList const Torrent::detailStatKeys{
     {
-        myValues[i].setValue(value);
-        changed = true;
+        setme = value;
     }
 
     return changed;
 }
 
-bool Torrent::setBool(int i, bool value)
+bool change(double& setme, double const& value)
 {
-    bool changed = false;
+    bool const changed = !qFuzzyCompare(setme + 1, value + 1);
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::Bool);
-
-    if (myValues[i].isNull() || myValues[i].toBool() != value)
+    if (changed)
     {
-        myValues[i].setValue(value);
-        changed = true;
+        setme = value;
     }
 
     return changed;
 }
 
-bool Torrent::setDouble(int i, double value)
+bool change(double& setme, tr_variant const* value)
 {
-    bool changed = false;
-
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::Double);
-
-    if (myValues[i] != value)
-    {
-        myValues[i].setValue(value);
-        changed = true;
-    }
-
-    return changed;
+    double d;
+    return tr_variantGetReal(value, &d) && change(setme, d);
 }
 
-bool Torrent::setTime(int i, time_t value)
+bool change(int& setme, tr_variant const* value)
 {
-    bool changed = false;
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<int>(i));
+}
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::DateTime);
+bool change(uint64_t& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<uint64_t>(i));
+}
 
-    auto& oldval = myValues[i];
-    auto const newval = qlonglong(value);
+bool change(time_t& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, static_cast<time_t>(i));
+}
 
-    if (oldval != newval)
-    {
-        oldval = newval;
-        changed = true;
-    }
+bool change(Speed& setme, tr_variant const* value)
+{
+    int64_t i;
+    return tr_variantGetInt(value, &i) && change(setme, Speed::fromBps(i));
+}
 
-    return changed;
+bool change(bool& setme, tr_variant const* value)
+{
+    bool b;
+    return tr_variantGetBool(value, &b) && change(setme, b);
 }
 
-bool Torrent::setSize(int i, qulonglong value)
+bool change(QString& setme, tr_variant const* value)
 {
     bool changed = false;
+    char const* str;
+    size_t len;
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::ULongLong);
+    if (!tr_variantGetStr(value, &str, &len))
+    {
+        return changed;
+    }
 
-    if (myValues[i].isNull() || myValues[i].toULongLong() != value)
+    if (len == 0)
     {
-        myValues[i].setValue(value);
-        changed = true;
+        changed = !setme.isEmpty();
+        setme.clear();
+        return changed;
     }
 
-    return changed;
+    return change(setme, QString::fromUtf8(str, len));
 }
 
-bool Torrent::setString(int i, char const* value, size_t len)
+bool change(Peer& setme, tr_variant* value)
```

Comment:
Here and elsewhere: consider making this const-correct, i.e. passing `tr_variant const* value` and doing `const_cast<tr_variant*>(value)` where necessary, as we know that the value doesn't in fact change.

Reply:
Changed in 3ddf787b7dd23954197cca2e7f8487d2da100e5e